File: Book Recommendation/R/ParseData.R
parse_book_crossing_data <- function(
  data_path='https://raw.githubusercontent.com/ChicagoBoothML/DATA___BookCrossing/master') {

  library(data.table)
  
  # Common NAs:
  na_strings <- c(
    '',
    'na', 'n.a', 'n.a.',
    'nan', 'n.a.n', 'n.a.n.',
    'NA', 'N.A', 'N.A.',
    'NaN', 'N.a.N', 'N.a.N.',
    'NAN', 'N.A.N', 'N.A.N.',
    'nil', 'Nil', 'NIL',
    'null', 'Null', 'NULL')
  
  cat('Parsing Books... ')
  books <- read.csv(
    file.path(data_path, 'BX-Books.csv'),
    sep=';',
    check.names=FALSE,
    colClasses='character',
    stringsAsFactors=FALSE,
    na.strings=na_strings)
  books <- as.data.table(
    books[ , c('ISBN', 'Book-Title', 'Book-Author')])
  names(books) <- c('isbn', 'book_title', 'book_author')
  cat('done!\n')
  
  cat('Parsing Users... ')
  users <- fread(
    file.path(data_path, 'BX-Users.csv'),
    sep=';',
    col.names=c('user_id', 'location', 'age'),
    colClasses='character',
    stringsAsFactors=FALSE,
    na.strings=na_strings,
    showProgress=FALSE)
  users$user_id <- as.integer(users$user_id)
  users$age <- as.numeric(users$age)
  cat('done!\n')
  
  cat('Parsing Ratings... ')
  ratings <- fread(
    file.path(data_path, 'BX-Book-Ratings.csv'),
    sep=';',
    col.names=c('user_id', 'isbn', 'book_rating'), 
    colClasses='character',
    stringsAsFactors=FALSE,
    na.strings=na_strings,
    showProgress=FALSE)
  ratings$user_id <- as.integer(ratings$user_id)
  ratings$book_rating <- as.numeric(ratings$book_rating)
  cat('done!\n')
  
  list(books=books, users=users, ratings=ratings)
}

File: Book Recommendation/Python/ParseData.py
from pandas import read_csv
from os.path import join

from ChicagoBoothML_Helpy.Print import printflush


def parse_book_crossing_data(
        data_path='https://raw.githubusercontent.com/ChicagoBoothML/DATA___BookCrossing/master'):

    # Common NAs:
    na_strings = [
        '',
        'na', 'n.a', 'n.a.',
        'nan', 'n.a.n', 'n.a.n.',
        'NA', 'N.A', 'N.A.',
        'NaN', 'N.a.N', 'N.a.N.',
        'NAN', 'N.A.N', 'N.A.N.',
        'nil', 'Nil', 'NIL',
        'null', 'Null', 'NULL']

    printflush('Parsing Books...', end=' ')
    books = read_csv(
        join(data_path, 'BX-Books.csv'),
        sep=';',
        dtype=str,
        na_values=na_strings,
        usecols=['ISBN', 'Book-Title', 'Book-Author'],
        error_bad_lines=False)
    printflush('done!')

    printflush('Parsing Users...', end=' ')
    users = read_csv(
        join(data_path, 'BX-Users.csv'),
        sep=';',
        dtype=str,
        na_values=na_strings,
        error_bad_lines=False)
    users['User-ID'] = users['User-ID'].astype(int)
    users['Age'] = users['Age'].astype(float)
    printflush('done!')

    printflush('Parsing Ratings...', end=' ')
    ratings = read_csv(
        join(data_path, 'BX-Book-Ratings.csv'),
        sep=';',
        dtype=str,
        na_values=na_strings,
        error_bad_lines=False)
    ratings['User-ID'] = ratings['User-ID'].astype(int)
    ratings['Book-Rating'] = ratings['Book-Rating'].astype(float)
    printflush('done!')

    return dict(books=books, users=users, ratings=ratings)
